lib/tpm/validate.go

```diff
@@ -0,0 +1,217 @@
+	EKCertSerial string `json:"ek_cert_serial,omitempty"`
+	// EKCertVerified is true if the EKCert was verified against the allowed
+	// CAs.
+	EKCertVerified bool `json:"ek_cert_verified"`
+}
+
+// JoinAuditAttributes returns a series of attributes that can be inserted into
+// audit events related to a specific join.
+func (c *ValidatedTPM) JoinAuditAttributes() (map[string]interface{}, error) {
+	res := map[string]interface{}{}
+	d, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
+		TagName: "json",
+		Result:  &res,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := d.Decode(c); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return res, nil
+}
+
+// Validate takes the parameters from a remote TPM and performs the necessary
+// initial checks before then generating an encrypted credential challenge for
+// the client to solve in a credential activation ceremony. This allows us to
+// verify that the client possesses the TPM corresponding to the EK public key
+// or certificate presented by the client.
+func Validate(
+	ctx context.Context, log *slog.Logger, params ValidateParams,
+) (*ValidatedTPM, error) {
+	ctx, span := tracer.Start(ctx, "Validate")
+	defer span.End()
+
+	ekCert, ekPub, err := parseEK(ctx, params)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	validated := &ValidatedTPM{}
+	validated.EKPubHash, err = hashEKPub(ekPub)
+	if err != nil {
+		return validated, trace.Wrap(err, "hashing EK public key")
+	}
+	if ekCert != nil {
+		validated.EKCertSerial = serialString(ekCert.SerialNumber)
+	}
+
+	if len(params.AllowedCAs) > 0 {
+		if err := verifyEKCert(ctx, params.AllowedCAs, ekCert); err != nil {
+			log.ErrorContext(
+				ctx,
+				"EKCert CA verification failed",
+				"error", err,
+				"tpm", validated,
+			)
+			return validated, trace.Wrap(err)
+		}
+		validated.EKCertVerified = true
+	}
+
+	activationParameters := attest.ActivationParameters{
+		TPMVersion: attest.TPMVersion20,
+		AK:         params.AttestParams,
+		EK:         ekPub,
+	}
+	// The generate method completes initial validation that provides the
+	// following assurances:
+	// - The attestation key is of a secure length
+	// - The attestation key is marked as created within a TPM
+	// - The attestation key is marked as restricted (e.g cannot be used to
+	//   sign or decrypt external data)
+	// When the returned challenge is solved by the TPM using ActivateCredential
+	// the following additional assurance is given:
+	// - The attestation key resides in the same TPM as the endorsement key
+	solution, encryptedCredential, err := activationParameters.Generate()
+	if err != nil {
+		return validated, trace.Wrap(err, "generating credential activation challenge")
+	}
+	clientSolution, err := params.Solve(encryptedCredential)
+	if err != nil {
+		return validated, trace.Wrap(err, "asking client to perform credential activation")
+	}
+	if subtle.ConstantTimeCompare(clientSolution, solution) == 0 {
+		log.ErrorContext(
```

Comment:
The problem with just returning an error is that the Teleport API does not have a middleware which automatically logs errors, and failed joins do not currently show up in the audit log. I chose to log here because under no usual circumstances should this fail, and a failure would likely be caused by a bad actor.

That being said, as I was working on this, I've also raised https://github.com/gravitational/teleport/pull/40329 which adds log messages and audit logs for join failures. So I could probably remove these log points from here now.

Reply:
Dropped these in favour of #40329